File: 410-split-array-largest-sum/split-array-largest-sum.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */

var splitArray = function (nums, k) {
    let max = Math.max(...nums);
    let sum = 0;
    let n = nums.length

    for (let i = 0; i < n; i++) {
        sum += nums[i]
    }

    function check(max, arr, k) {
        let sub = 1;
        let sum = 0;

        for (let i = 0; i < n; i++) {
            if (arr[i] + sum <= max) {
                sum += arr[i]
            } else {
                sub++;
                sum = arr[i];
            }
        }


        return sub
    }


    let low = max;
    let high = sum;

    let ans = -1;
    while (low <= high) {
        let mid = Math.floor((low + high) / 2);

        if (check(mid, nums, k) <= k) {
            ans = mid;
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }

    return low;
};

File: 410-split-array-largest-sum/split-array-largest-sum.java
class Solution {
  int check(int maxSum, int k, int[] arr) {
    int sum = 0;
    int cnt = 1;

    for (int num : arr) {
      if (sum + num <= maxSum) {
        sum += num;
      } else {
        cnt++;
        sum = num;
      }
    }

    return cnt;
  }

  public int splitArray(int[] nums, int k) {
    int max = nums[0];
    int sum = 0;

    for (int num : nums) {
      sum += num;
      max = Math.max(max, num);
    }

    int low = max;
    int high = sum;
    int res = 0;

    while (low <= high) {
      int mid = (low + high) / 2;

      if (check(mid, k, nums) <= k) {
        res = mid;
        high = mid - 1;
      } else {
        low = mid + 1;
      }
    }

    return res;
  }
}